### PR TITLE
[FilterControl] Adds placeholder prop

### LINKED ISSUE #1256: ResourceList placeholder prop
# Feature request summary
Adding a `placeholder` prop to the `ResourceList` to allow users to set their own messages in the search area.

## Rationale

image

Adding a prop like this would allow users to customize the placeholder text shown in the search area. Right now it defaults to the value in the `resourceName` prop.

---

🌟 Feature requests that are not yet planned will be closed. We then use the issue’s [:+1: upvotes](https://help.github.com/articles/about-conversations-on-github/) to track and set priorities. See the [contribution guidelines](https://github.com/Shopify/polaris-react/blob/master/.github/CONTRIBUTING.md) for more information.

### PR DESCRIPTION
**Pull request by Rchristiani**
### WHY are these changes introduced?

Fixes #1256 

Currently the `FilterControl` component will use the `resourceName.plural` as the placeholder to the `TextField`. This PR adds a `placeholder` prop to `FilterControl` to allow users to have a custom placeholder.

### WHAT is this pull request doing?

I added an optional `placeholder` prop to the `FilterControl` component. If passed the `placeholder` on the `TextField` will use the value provided, if not passed it will default to the `resourceName.plural` value.

I also removed a `,` from the `mockDefaultContext` plural string, as it seemed unused. Correct me if i am wrong!

I didn't test accessibility since it is passing a string to the `TextField` so I don't think that is required?

### How to 🎩

🖥 [Local development instructions](https://github.com/Shopify/polaris-react/blob/master/README.md#development)
🗒 [General tophatting guidelines](https://github.com/Shopify/polaris-react/blob/master/documentation/Tophatting.md)
📄 [Changelog guidelines](https://github.com/Shopify/polaris-react/blob/master/documentation/Versioning%20and%20changelog.md)

<details>
<summary>Copy-paste this code in <code>playground/Playground.tsx</code>:</summary>

```jsx
import * as React from 'react';
import {Page, ResourceList, Avatar, TextStyle, FilterType, Card} from '../src';

interface State {}

class ResourceListExample extends React.Component {
  state = {
    searchValue: '',
  };

  handleSearchChange = (searchValue) => {
    this.setState({searchValue});
  };

  renderItem = (item) => {
    const {id, url, name, location} = item;
    const media = <Avatar customer size="medium" name={name} />;

    return (
      <ResourceList.Item id={id} url={url} media={media}>
        <h3>
          <TextStyle variation="strong">{name}</TextStyle>
        </h3>
        <div>{location}</div>
      </ResourceList.Item>
    );
  };

  render() {
    const resourceName = {
      singular: 'customer',
      plural: 'customers',
    };

    const items = [
      {
        id: 341,
        url: 'customers/341',
        name: 'Mae Jemison',
        location: 'Decatur, USA',
      },
    ];

    const filterControl = (
      <ResourceList.FilterControl
        placeholder="by name, email or phone"
        searchValue={this.state.searchValue}
        onSearchChange={this.handleSearchChange}
        additionalAction={{
          content: 'Save',
          onAction: () => console.log('New filter saved'),
        }}
      />
    );

    return (
      <Card>
        <ResourceList
          resourceName={resourceName}
          items={items}
          renderItem={this.renderItem}
          filterControl={filterControl}
        />
      </Card>
    );
  }
}

export default class Playground extends React.Component<{}, State> {
  render() {
    return (
      <Page title="Playground">
        {/* Add the code you want to test in here */}
        <ResourceListExample />
      </Page>
    );
  }
}
```

</details>

### 🎩 checklist

* [x] Tested on [mobile](https://github.com/Shopify/polaris-react/blob/master/documentation/Tophatting.md#cross-browser-testing)
* [x] Tested on [multiple browsers](https://help.shopify.com/en/manual/intro-to-shopify/shopify-admin/supported-browsers)
* [ ] Tested for [accessibility](https://github.com/Shopify/polaris-react/blob/master/documentation/Accessibility%20testing.md)

### CODE REVIEW

File: src/components/ResourceList/components/FilterControl/tests/FilterControl.test.tsx [hunk 1]
```diff
@@ -22,7 +22,7 @@ describe('<FilterControl />', () => {
     selectMode: false,
     resourceName: {
       singular: 'item',
-      plural: 'items,',
+      plural: 'items',
```
Comment (dleroux) [inline src/components/ResourceList/components/FilterControl/tests/FilterControl.test.tsx]: why did you remove the `s`

File: src/components/ResourceList/components/FilterControl/FilterControl.tsx [hunk 2]
```diff
@@ -53,6 +55,12 @@ export class FilterControl extends React.Component<CombinedProps> {
       },
     );
 
+    const placeholderLabel = placeholder
```
Comment (dleroux) [inline src/components/ResourceList/components/FilterControl/FilterControl.tsx]: I would think it's probably best that when a custom placeHolderLabel is provided, the consumer of in responsible for passing the translation.

```
const textFieldLabel = placeholder 
 ? placeholder 
 : intl.translate(
      'Polaris.ResourceList.FilterControl.textFieldLabel',
      {
        resourceNamePlural: resourceName.plural.toLocaleLowerCase(),
      },
    );
```

File: src/components/ResourceList/components/FilterControl/tests/FilterControl.test.tsx [hunk 3]
```diff
@@ -637,6 +637,37 @@ describe('<FilterControl />', () => {
...
+  describe('placeholder', () => {
+    it('renders default text if no placeholder passed.', () => {
+      const filterControl = mountWithAppProvider(
+        <Provider value={mockDefaultContext}>
+          <FilterControl {...mockDefaultProps} filters={mockFilters} />
+        </Provider>,
+      );
+
+      expect(filterControl.find(TextField).prop('placeholder')).toBe(
+        'Search items',
```
Comment (dleroux) [inline src/components/ResourceList/components/FilterControl/tests/FilterControl.test.tsx]: You'll need to translate that string, but if we opt to not translate it and let the consumer pass the translate text then we just need to check that it was passed as is.

File: src/components/ResourceList/components/FilterControl/tests/FilterControl.test.tsx [hunk 4]
```diff
@@ -637,6 +637,37 @@ describe('<FilterControl />', () => {
...
+      );
+    });
+
+    it('renders the placeholder prop value if provided', () => {
+      const placeholder = 'by name, email or phone';
```
Comment (dleroux) [inline src/components/ResourceList/components/FilterControl/tests/FilterControl.test.tsx]: See this does makes me think it's better to be 100% up to the user what they pass in. Search... was opinionated base we knew what we would render, but having to provide a placeholder that will be prepended by Search requires the consumer to know this.

File: src/components/ResourceList/components/FilterControl/tests/FilterControl.test.tsx (hunk 1 above)
Comment (Rchristiani) [reply src/components/ResourceList/components/FilterControl/tests/FilterControl.test.tsx]: I didn't remove the `s` it was a `,` that seemed to have no tests against it. So when I was writing my test and checking the default the value was `Search items,` and that made no sense to me.

File: src/components/ResourceList/components/FilterControl/FilterControl.tsx (hunk 2 above)
Comment (Rchristiani) [reply src/components/ResourceList/components/FilterControl/FilterControl.tsx]: Are you thinking that instead of translating the `Search ` part we can just leave the transitional completely up to the consumer? That sounds like a good idea!

File: src/components/ResourceList/components/FilterControl/tests/FilterControl.test.tsx (hunk 3 above)
Comment (Rchristiani) [reply src/components/ResourceList/components/FilterControl/tests/FilterControl.test.tsx]: I think if we go with the consumer doing the translation, then I am cool just testing the value passed as is.

File: src/components/ResourceList/components/FilterControl/tests/FilterControl.test.tsx (hunk 4 above)
Comment (Rchristiani) [reply src/components/ResourceList/components/FilterControl/tests/FilterControl.test.tsx]: OK, sounds good! So just to confirm I will remove the `Search` part and the `placeholder` can be anything the user wants it to be?

File: src/components/ResourceList/components/FilterControl/FilterControl.tsx (hunk 2 above)
Comment (dleroux) [reply src/components/ResourceList/components/FilterControl/FilterControl.tsx]: Yeah, either they use the default we have today, nothing gets changed there but if they do provide a placeholder text it gets used for both the label and the placeholder (because in thsi case they would be the same) as provided by the consumer.

File: src/components/ResourceList/components/FilterControl/tests/FilterControl.test.tsx (hunk 4 above)
Comment (dleroux) [reply src/components/ResourceList/components/FilterControl/tests/FilterControl.test.tsx]: Yeah I would think so.

File: src/locales/en.json [hunk 5]
```diff
@@ -171,7 +171,8 @@
       },
 
       "FilterControl": {
-        "textFieldLabel": "Search {resourceNamePlural}"
+        "textFieldLabel": "Search {resourceNamePlural}",
+        "placeholderLabel": "Search {placeholder}"
```
Comment (dleroux) [inline src/locales/en.json]: 🔥
